### PR TITLE
Support exception tests for java.

### PR DESCRIPTION
**Pull request by shinfan**

### CODE REVIEW

File: src/main/resources/com/google/api/codegen/java/mock_service_impl.snip [hunk 1]
```diff
@@ -52,23 +56,35 @@
...
-    responseObserver.onCompleted();
+    Object response = responses.remove();
+    if (response instanceof {@method.responseTypeName}) {
+      requests.add(request);
+      responseObserver.onNext(({@method.responseTypeName}) response);
+      responseObserver.onCompleted();
+    } else if(response instanceof Exception) {
+      responseObserver.onError((Exception) response);
+    } else {
+      throw new IllegalArgumentException("Unrecognized response type");
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/mock_service_impl.snip]: How does this surface when it happens?

File: src/main/resources/com/google/api/codegen/java/test.snip [hunk 2]
```diff
@@ -80,6 +80,24 @@
         actualRequest.{@assert.actualValueGetter}());
     @end
   }
+
+  @@Test
+  @@SuppressWarnings("all")
+  public void {@test.name}Exception() {
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/test.snip]: Don't concatenate strings in the templates!

File: src/main/resources/com/google/api/codegen/java/test.snip [hunk 3]
```diff
@@ -80,6 +80,24 @@
...
+    try {
+      {@initCode(test.initCode)}
+
+      api.{@test.surfaceMethodName}(\
+          {@sampleMethodCallArgList(test.initCode.fieldSettings)});
+      Assert.fail("No exception raised");
+    } catch (ApiException e) {
+      Assert.assertEquals(Status.INTERNAL.getCode(), e.getStatusCode());
+    } catch (Exception e) {
+      Assert.fail("Unknown exception found");
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/test.snip]: We could remove this final catch, and just have the test declare `throws Exception`.

File: src/main/resources/com/google/api/codegen/java/test.snip [hunk 4]
```diff
@@ -89,20 +107,53 @@
...
+
+@private streamingMethodCall(test)
+  @switch test.grpcStreamingType
+  @case "BidiStreaming"
+    {@bidiStreamingCall(test)}
+    requestObserver.onCompleted();
+  @case "ServerStreaming"
+    {@serverStreamingCall(test)}
+  @default
+    $unhandled case - test.grpcStreamingType.toString$
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/test.snip]: To be consistent, use the prefix `$unhandledCase: ` (no dash)

File: src/main/resources/com/google/api/codegen/java/mock_service_impl.snip [hunk 5]
```diff
@@ -52,23 +56,35 @@
...
-    final {@method.responseTypeName} response = ({@method.responseTypeName}) responses.remove();
+    final Object response = responses.remove();
     StreamObserver<{@method.requestTypeName}> requestObserver =
         new StreamObserver<{@method.requestTypeName}>() {
       @@Override
       public void onNext({@method.requestTypeName} value) {
-        responseObserver.onNext(response);
+        if (response instanceof {@method.responseTypeName}) {
+          responseObserver.onNext(({@method.responseTypeName}) response);
+        } else if(response instanceof Exception) {
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/java/mock_service_impl.snip]: add space after `if`

Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/mock_service_impl.snip]: Done

File: src/main/resources/com/google/api/codegen/java/test.snip (hunk 2 above)
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/test.snip]: Done

File: src/main/resources/com/google/api/codegen/java/test.snip (hunk 4 above)
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/test.snip]: Done

File: src/main/resources/com/google/api/codegen/java/mock_service_impl.snip (hunk 1 above)
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/java/mock_service_impl.snip]: I switched to use` responseObserver.onError()` so an exception would be raised in the corresponding test.

**Comment by shinfan** [comment]
PTAL

**Comment by garrettjonesgoogle** [comment]
LGTM

**Pull request merged**